memtier: more regular annotation interpretation for CPU allocation preferences.

klihub opened the pull request:
This PR is based on/requires PR #588.

Update the interpretation of shared/isolated CPU opt-in and opt-out annotations
(and global policy defaults) to produce less irregular and hopefully more intuitive
CPU allocation preferences.

TODO/Question:
Would it provide better usability to use a single explicit non-boolean annotation
(for instance `cpu-preferences`) to express all CPU allocation preferences ?

askervin commented on pkg/cri/resource-manager/policy/builtin/memtier/pod-preferences.go:
```diff
@@ -217,62 +325,123 @@ func coldStartPreference(pod cache.Pod, container cache.Container) (ColdStartPre
-			} else {
-				isolate = true
+	// sub-core CPU request
+	case cores == 0 && fraction < 1000:
```
`fraction < 1000` is always true after line 392 (`fraction = fraction % 1000`)

klihub replied:
Thanks, true. Removed tautological comparison.

askervin commented on pkg/cri/resource-manager/policy/builtin/memtier/pod-preferences.go:
```diff
@@ -217,62 +325,123 @@ func coldStartPreference(pod cache.Pod, container cache.Container) (ColdStartPre
 			}
-		case full > 1:
-			isolate = preferIsol && explicit
+			return cores, fraction, preferIsolated, cpuNormal
```
This may be just because of lacking some background information, but I find it somewhat difficult to understand/explain/document why scheduling on isolated cores requires explicit annotation when requesting 2100m CPU but not when requesting 1900m CPU.

Do you think we could simplify the behavior by using `preferIsolated && explicitIsolated` in every case, or am I missing some important use case?

marquiz replied:
@klihub can correct me if I'm wrong, but, I guess the thinking is that allocating exactly one isolated core is "safe" from the workload scheduling point of view. When/if allocating more than one isolated cores the workload must be "isolcpus aware" and handle scheduling/cpu affinity of its threads by itself. In the case of one core this is not a problem as all threads will be forced on this one core

klihub replied:
Yes, it's exactly as @marquiz said. And that's why we don't assume all workloads to be safe for multi-core mixed allocations but require them to be annotated so.

marquiz replied:
K. Agree with @askervin that this should be documented. A comment in the source code would probably be in order

klihub replied:
I added a rationale section to the comments explaining the reasons behind the choices when to allocate
isolated vs. ordinary cores in the various cases.
Fixed a comparison thinko in the comments and a typo.
Removed a useless ` + 1000 * fraction` from a `return` on a branch where `fraction` was guaranteed to
always be `0`.

askervin reviewed (commented):
Looks good already.

Added a nitpick and a question on isolated cores to understand why the logic is like this...

askervin commented:
> Would it provide better usability to use a single explicit non-boolean annotation
(for instance cpu-preferences) to express all CPU allocation preferences ?

Elaborating this a bit... At minimum current annotations would be replaced by:

`cpu-preferences.cri-resource-manager.intel.com/pod: shared`
`cpu-preferences.cri-resource-manager.intel.com/pod: isolated`

...and if we'd like to enable running a `kube-system` pod on exclusive CPUs (instead of running it on always-shared reserved CPUs, this might be useful), or a non-kube-system pod on reserved CPUs, I could imagine these to make sense as well:

`cpu-preferences.cri-resource-manager.intel.com/pod: normal`
`cpu-preferences.cri-resource-manager.intel.com/pod: reserved`

Some combinations would be valid and useful:

`cpu-preferences.cri-resource-manager.intel.com/pod: normal,shared`
`cpu-preferences.cri-resource-manager.intel.com/pod: normal,isolated`

Some combinations would not be supported:

`cpu-preferences.cri-resource-manager.intel.com/pod: reserved,isolated`

Some combinations would be contradictory:

`cpu-preferences.cri-resource-manager.intel.com/pod: shared,isolated`
`cpu-preferences.cri-resource-manager.intel.com/pod: normal,reserved`

Did you have something like this in mind? (I chose these words and combinations directly based on the implementation, which doesn't guarantee perfect user experience...)

klihub commented:
> > Would it provide better usability to use a single explicit non-boolean annotation
> > (for instance cpu-preferences) to express all CPU allocation preferences ?
> 
> Elaborating this a bit... At minimum current annotations would be replaced by:
> 
...
> Did you have something like this in mind? (I chose these words and combinations directly based on the implementation, which doesn't guarantee perfect user experience...)

Yes, something like that. A set of allocation preferences with further possible qualifiers. Eg. `{ shared, exclusive, mixed }`, and `{isolated, non-isolated}`, although I'd probably throw in EPP/SS-related ones as well to allow explicitly expressed opt-in/opt-out preferences for CPUs in different performance profile classes.

There are a few clear benefits of using a single annotation. For instance, it is obvious that you're tweaking inter-related things, and in many cases it could be more intuitive/descriptive what you want/mean and easier to understand than with a set of implicitly inter-related booleans. But as you said, it is not completely without problems.

askervin reviewed (approved):
Nice comment improvement!

LGTM, works great on top of current master.

kad commented:
my line of thoughts: those so far are "preferred". what about "required" ?
Also, if we change some treatments of annotations - that should be reflected in the documentation as well.

klihub commented:
> my line of thoughts: those so far are "preferred". what about "required" ?

Wouldn't the easiest/most natural way to handle required vs. preferred CPU allocation preferences be the above outlined single, combined annotation, and qualifiers to denote the preferred vs. mandatory nature of the request ?

> Also, if we change some treatments of annotations - that should be reflected in the documentation as well.

The exact interpretation of these annotations was never properly documented in detail, and I agree that we
need to document it.

kad commented:
> > Also, if we change some treatments of annotations - that should be reflected in the documentation as well.
> 
> The exact interpretation of these annotations was never properly documented in detail, and I agree that we
> need to document it.

My comment was more about general practice: if we change some behaviour, and while documenting it in code comments is ok, we need to change userfacing doc in the same PR, so it wouldn't be lost.

klihub commented:
> My comment was more about general practice: if we change some behaviour, and while documenting it in code comments is ok, we need to change userfacing doc in the same PR, so it wouldn't be lost.

Ah, ok. AFAICT, this PR does not change documented behavior. Partly this is because the behavior was
never properly documented in the first place. But in reality, mostly because it was not the intention. This
PR is about spelling out saner behavior for some less used/obvious edge cases, and in a more explicit
manner.

Anyway, I wrote an initial version of a more proper memtier description. It's now at the tip of this PR. I'm
fine with including it as part of this PR. IMO, ideally it should be a separate PR. @kad, @marquiz,
@askervin, let me know guys what is the consensus preference and I'll adjust my PR accordingly.

marquiz reviewed (commented):
I've really only looked at the documentation which is a huge improvement over the previous. Really good job 👍  One comment there is that I'd add some references (links) to some terms like "PMEM", "HBM", "2-layer mode", "isolcpus" for example. WDYT?

askervin commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+The memtier policy has the following features:
+
+  - topologically aligned allocation of CPU and memory
+    * assign CPU and memory to a workloads with tightest available alignment
```
"a workloads" -> workloads

klihub replied:
Fixed.

askervin commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+  - topologically aligned allocation of CPU and memory
+    * assign CPU and memory to a workloads with tightest available alignment
+  - aligned allocation of devices
+    * pick pool for workload based on localoty of devices already assigned
```
"localoty" -> "locality"

klihub replied:
Fixed.

askervin commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+  - shared allocation of CPU cores
+    * assign workload to shared subset of pool CPUs
+  - exclusive allocation of CPU cores
+    * slice off CPU cores from shared subset and assign to workload
```
"slice off"
->
"dynamically slice off"

This definitely is a matter of taste... but I'd emphasize that the size of the set of shared CPUs in a pool indeed changes during runtime. This is pretty cool feature.

klihub replied:
That is mentioned/explained elsewhere in the text: "This shared subset can and usually does change
dynamically as exclusive CPU cores are allocated and released in the pool." Nevertheless, I updated feature list entry as suggested.

askervin commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+    * slice off CPU cores from shared subset and assign to workload
+  - mixed allocation of CPU cores
+    * assign both exclusive and shared CPU cores to workload
+  - discover and use kernel-isolated CPU cores
```
"discover and use" -> "discovering and using"

klihub replied:
Updated.

askervin commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+  - `kube-system` group
+    * all containers in the `kube-system` namespace
+  - `low-priority` group
+    * containers in the `BestEffort or Burstable` QoS class
```
`BestEffort or Burstable`
->
 `BestEffort` or `Burstable`

klihub replied:
Updated.

askervin commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+    * not eligible for either if annotated to opt out from exclusive allocation
+    * not eligible for isolated allocation if annotated to opt out
+  - `multi-core`
+    * CPU request fractional (`(CPU request % 1000 milli-CPU) != CPU request`):
```
`(CPU request % 1000 milli-CPU) != CPU request`
->
`(CPU request % 1000 milli-CPU) != 0`
or
`int(CPU request / 1000 milli-CPU) != CPU request`
or
(`CPU request` is not divisible by 1000 milli-CPU)

klihub replied:
Fixed.

askervin commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+which underdeclares `ReservedResources`. In that case, ordinary shared CPU cores
+will be used instead of kube-reserved ones.
+
+Eligibility for exclusive CPU allocation should always be possible to honor. If
```
"Eligibility for exclusive CPU allocation should always be possible to honor. ..."

I wouldn't bother to mention the possibility of a bug in the algorithm. There shouldn't be one, and on the other hand, if there is, it could break any other claim as well, not just this.

However, a configuration error like too large ReservedResources cpuset could cause problems. That takes CPUs from real pools without kube scheduler knowing about it.

klihub replied:
Removed reference to possible bugs in resource accounting.

askervin commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
 
-The `memtier` policy will then aim to allocate resources from a topology
-node which can satisfy the memory requirements.
+While this might change in the future, currently implicit topology hint
```
"While this might change in the future, currently implicit topology hint..."
->
"Topology hint..."

(I'd prefer documenting only the current behavior and avoid speculating on what could change... unless we already know for sure something is going to change soon.)

klihub replied:
Removed.

askervin commented on docs/policy/memtier.md:
````diff
@@ -1,70 +1,314 @@
+```yaml
+metadata:
+  annotations:
+    # opt in container $C1 to shared CPU core allocation
````
In existing examples we don't use `$` as part of container names. Perhaps we could go here with `$`, too? (At least it made me wonder if one needs to prefix container names by `$` when using this notation.)

klihub replied:
Removed $ from all example container names.

askervin commented:
Great stuff! Didn't notice any big gap, just tiny fixes.

klihub commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+memory zones in the socket's dies, and the root ends up containing all CPU
+cores and memory zones in all sockets. The actual pool tree is built with all
+non-branching paths eliminated, by omitting any child node which would have
+ended up being the lone single child of its parent. This is merely an minor
```
Fixed.

klihub commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+workload can vary greatly, depending on how efficiently data is transferred
+between them or, in other words, on how well the resources are aligned.
+
+There are a number of inherent architectural properties of server grade
```
Fixed.

klihub commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+a gradually increasing maximum potential cost or penalty for memory access and
+data transfer between CPU cores.
+
+Another property of this setup is that the resources sets of sibling pools at
```
Fixed.

klihub commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+  - cold start
+    * pin workload exclusively to PMEM for an initial warm-up period
+  - dynamic page demotion
+    * forcibly migrate read-only and idle container memory pages to PMEM
```
It's there: "- discover and use kernel-isolated CPU cores"...

klihub replied:
Added explicit reference to 'isolcpus'.

klihub commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+bandwidth to each memory controller and I/O device.
+
+If a workload is not assigned to run with a properly aligned set of CPU,
+memory, and devices, it will not be able to achieve optimal performance.
```
Fixed.

klihub commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+### Shared, Exclusive, and Isolated CPU Preference
+
+A container can opt in to or opt out from shared CPU allocation using the
+following annotation.
```
I replaced all references to 'annotations' immediately preceding examples with 'Pod annotations'.

klihub commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+container's assigned pool. This shared subset can and usually does change
+dynamically as exclusive CPU cores are allocated and released in the pool.
+
+## Container CPU Allocation Preference Annotations
+
+Containers can be annotated to diverge from the default CPU allocation
+preferences the policy would otherwise apply to them. These annotations
+can be given both with per pod and per container resolution. If for any
+container both of these exist, the container-specific one takes precedence.
```
I replaced most references to 'annotations' with 'Pod annotations'.

klihub commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+the workload and then assigning CPU and memory from this pool.
+
+The pool nodes at various depths from bottom to top represent the NUMA nodes,
+dies, sockets, and finally the whole of the system a the root node. Leaf NUMA
```
Fixed.

klihub commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+of memory is assigned to the closest NUMA node pool.
+
+Each non-leaf pool node in the tree is assigned the union of the resources of
+its children. So in practice, dies end up containing the closest CPU cores
```
Reworded and eliminated 'closest' from the text.

klihub commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
+non-branching paths eliminated, by omitting any child node which would have
+ended up being the lone single child of its parent. This is merely an minor
+optimisation in the implementation and it has no practical significance for
+this discussion.
```
Removed the last few irrelevant sentences that state the slight divergence of the implementation from the preceding description due to a small practical optimization.

klihub commented on docs/policy/memtier.md:
```diff
@@ -1,70 +1,314 @@
 metadata:
   annotations:
-    memory-type.cri-resource-manager.intel.com/container.container1: dram,pmem
+    # opt in container $C1 to isolted exclusive CPU core allocation
```
Fixed.

klihub commented:
> I've really only looked at the documentation which is a huge improvement over the previous. Really good job  One comment there is that I'd add some references (links) to some terms like "PMEM", "HBM", "2-layer mode", "isolcpus" for example. WDYT?

Good idea. I'll try to find appropriate online content to link to (it's easy for isolcpus) and add links.

klihub commented:
> > I've really only looked at the documentation which is a huge improvement over the previous. Really good job  One comment there is that I'd add some references (links) to some terms like "PMEM", "HBM", "2-layer mode", "isolcpus" for example. WDYT?
> 
> Good idea. I'll try to find appropriate online content to link to (it's easy for isolcpus) and add links.

Okay, added the following external links as explanation:
  - isolcpus: https://www.kernel.org/doc/html/latest/admin-guide/kernel-parameters.html#cpu-lists
  - PMEM: https://www.intel.com/content/www/us/en/products/memory-storage/optane-dc-persistent-memory.html
  - HBM: https://en.wikipedia.org/wiki/High_Bandwidth_Memory

Any suggestions to better docs are welcome.

I also removed any references to '2-layer mode'. It was anyway incorrect as it
should have been 'two level memory' in the first place. However, since
1LM, 2LM, Memory Mode and App Direct Mode are just prone to create
more confusion than clarity, I opted for a short verbal clarification of what
'tiering' is in the scope of 'memtier'.

marquiz reviewed (approved):
lgtm
